### ISSUE TITLE
the hierarchical clustering result is not consistent with R and python

### ISSUE THREAD

**Issue by jingpengw** [JuliaStats/Clustering.jl#134]
It seems that Julia clustering result is different with R and python, here is my comparison based on my data:
![image](https://user-images.githubusercontent.com/7651573/49386620-f7361980-f6ed-11e8-8633-43757ad67890.png)

some reproducible code
the julia result
```
using Random

Random.seed!(123)

N = 8
condensedDistanceMatrix = rand(div(N*(N-1),2))

# construct distance matrix
distanceMatrix = zeros(Float64, N,N)

k = 0
for i in 1:N
    for j in i+1:N
        k += 1
        distanceMatrix[i,j] = condensedDistanceMatrix[k]
        distanceMatrix[j,i] = distanceMatrix[i,j]
    end
end

hc = hclust(distanceMatrix, linkage=:ward)

using Pkg
include(joinpath(Pkg.devdir(), "RealNeuralNetworks/src/Utils/PlotRecipes.jl"))
using .PlotRecipes
PlotRecipes.plot(hc)
```
![image](https://user-images.githubusercontent.com/7651573/49386433-842ca300-f6ed-11e8-8f8b-6fb7ce2007dc.png)

the python result
```
@pyimport scipy.cluster.hierarchy as pyhierarchy

Z = pyhierarchy.ward(condensedDistanceMatrix)

dn = pyhierarchy.dendrogram(Z)
```
![image](https://user-images.githubusercontent.com/7651573/49386387-724b0000-f6ed-11e8-98a1-2bce197d8a91.png)

it is very similar with squared clustering:

```
@pyimport scipy.cluster.hierarchy as pyhierarchy

Z = pyhierarchy.ward(condensedDistanceMatrix .* condensedDistanceMatrix)

dn = pyhierarchy.dendrogram(Z)
```
![image](https://user-images.githubusercontent.com/7651573/49386550-cce45c00-f6ed-11e8-8881-9e0be0a9794f.png)

the plot script `PlotRecipes.jl` I am using:
```julia
module PlotRecipes
using RealNeuralNetworks.Neurons
using Colors, ColorSchemes, Clustering
using Plots
using Statistics 

import PyPlot
PyPlot.svg(true)

function plot_synapse_distributions( cellList;  
        synapseDistribution = map(get_synapse_to_soma_path_length_lists, cellList),
        saveName = joinpath(homedir(), "synapse_distribution.svg"),
        prerange=(0,50), postrange=(0,100), bins=30)
    fig = PyPlot.figure()
    for (i,cellId) in enumerate(cellList)
        @show i
        fig[:add_subplot](4,2,(i-1)*2+1)
        PyPlot.plt[:hist](synapseDistribution[i][1]./1000; range=prerange, bins=bins)
        PyPlot.plt[:xlabel]("distrance from presynapse to soma ($cellId)")

        fig[:add_subplot](4,2,(i-1)*2+2)
        PyPlot.plt[:hist](synapseDistribution[i][2]./1000; range=postrange, bins=bins)
        PyPlot.plt[:xlabel]("distrance from postsynapse to soma ($cellId)")
        
    end 
    PyPlot.savefig(saveName)
end

function reorder_distance_matrix(distanceMatrix::Array, clust::Hclust)
	# rearrange distance matrix according to order
	reorderedDistanceMatrix = zeros(distanceMatrix)
	for (i,o1) in enumerate(clust.order)
		for (j, o2) in enumerate(clust.order)
			reorderedDistanceMatrix[i,j] = distanceMatrix[o1,o2]
		end 
	end 
	reorderedDistanceMatrix
end 

function coloring(dm)
    img = Array{RGB{Float64}}(size(dm))
    for i in eachindex(dm)
        img[i] = get(ColorSchemes.jet, dm[i])
    end 
    img
end 


function plot(neuron::Neuron; nodeStep::Integer=10)
    #using PyPlot
    PyPlot.pygui(true)

    # plot soma
    root = Neurons.get_root_node(neuron)
    PyPlot.scatter3D([root[1]], [root[2]], [root[3]])
        
    for segment in neuron
        nodeList = Segments.get_node_list(segment)
        x = map(n->n[1], nodeList[1:nodeStep:end])
        y = map(n->n[2], nodeList[1:nodeStep:end])
        z = map(n->n[3], nodeList[1:nodeStep:end])
        PyPlot.plot(x,y,z)
    end
end 

function plot_v1(neuron::Neuron; nodeStep::Integer=10)
    segmentList = neuron.segmentList
    plotly()
    for branch in segmentList
        nodeList = Neurons.Segments.get_node_list(branch)
        #@show length(nodeList)
        #@show nodeList
        x = map(n->n[1], nodeList[1:nodeStep:end])
        y = map(n->n[2], nodeList[1:nodeStep:end])
        z = map(n->n[3], nodeList[1:nodeStep:end])
        plot!(x,y,z) #color=rand(Colors.RGB))
    end 

    root = Neurons.get_root_node(neuron)
    plot!([root[1]], [root[2]], [root[3]], m=(2, :circle), leg=false)
end 

"""
    plot_maximum_intensity_projection(vol::Array; colorMap="gray")
color map could be "jet" if you like colorful display. 
"""
function plot_maximum_intensity_projection(vol::Array{T,3}; colorMap="gray") where T
    fig = PyPlot.figure()
    fig[:add_subplot](2,2,1)
    xy = maximum(vol, 3)[:,:,1]
    PyPlot.imshow(xy, colorMap)
    # colorbar()
    fig[:add_subplot](2,2,2)
    xz = maximum(vol, 2)[:,1,:]
    PyPlot.imshow(xz, colorMap)
    fig[:add_subplot](2,2,3)
    yz = maximum(vol, 1)[1,:,:] |> rotl90
    PyPlot.imshow(yz, colorMap)
    PyPlot.colorbar()
end 

function indexmap(x::Vector)
    ret = Dict()
    for (i,v) in enumerate(x)
        ret[v] = i
    end 
    ret
end

function hclustplot(hc::Hclust, useheight::Bool)
    o = indexmap(hc.order)
    n = [x for x in 1:length(o)]

    pos = treepositions(hc, useheight)


    xs = []
    ys = []
    for i in 1: size(hc.merges, 1)
        x1 = pos[hc.merges[i,1]][1]
        x2 = pos[hc.merges[i,2]][1]
        append!(xs, [x1,x1,x2,x2])

        y1 = pos[hc.merges[i,1]][2]
        y2 = pos[hc.merges[i,2]][2]
        useheight ? h = hc.heights[i] : h = 1
        newy = maximum([y1,y2]) + h
        append!(ys, [y1,newy,newy,y2])
    end
    return (reshape(xs, 4, size(hc.merges, 1)), reshape(ys, 4, size(hc.merges, 1)))
end

function treepositions(hc::Hclust, useheight::Bool)
    order = indexmap(hc.order)
    positions = Dict{}()
    for (k,v) in order
        positions[-k] = (v, 0)
    end
    for i in 1:size(hc.merges,1)
        xpos = mean([positions[hc.merges[i,1]][1], positions[hc.merges[i,2]][1]])
        if hc.merges[i,1] < 0 && hc.merges[i,2] < 0
            useheight ? ypos = hc.heights[i] : ypos = 1
        else
            useheight ? h = hc.heights[i] : h = 1
            ypos = maximum([positions[hc.merges[i,1]][2], positions[hc.merges[i,2]][2]]) + h
        end

        positions[i] = (xpos, ypos)
    end
    return positions
end 

function plot(clust::Hclust)
	Plots.gr()
	#Plots.plotlyjs()
	Plots.plot(hclustplot(clust, true), seriestype=:path, color=:black,
    grid=false, legend=false) #,  xticks=classificationIdList[clust.order])
end 

end # module
```

**Comment by alyst** [comment]
How do you check that it's not consistent? The trees look similar to me; I haven't looked into the plotting code you provide, but are you sure that the leaf labels in Julia plot are correct? For me it looks like X is just a numeric axis.
Actually, consistency with R results is quite extensively checked: see `tests/hclust-gen-examples.R`, `tests/hclust.jl` and `hclust_generated_examples.jl.gz` (we apply R's `hclust()` to randomly generated data and then compare it with Julia output for the same data).

**Comment by alyst** [comment]
Should I close the issue?

**Comment by diegozea** [comment]
The problem is in the heights that are indicated in the Julia plot. It looks like Julia reached higher heights than R and Python.

**Comment by diegozea** [comment]
Ok, I've just run an example from the [wiki](https://en.wikipedia.org/wiki/Complete-linkage_clustering) and the heights are OK. However, the `ytick` labels of the plot are wrong:

```
wiki =[
	0	17	21	31	23
	17	0	30	34	21
	21	30	0	28	39
	31	34	28	0	43
	23	21	39	43	0
]
```

![image](https://user-images.githubusercontent.com/2822757/88334895-bb245200-cd32-11ea-9bff-c665901733ae.png)

**Comment by diegozea** [comment]
PR https://github.com/JuliaPlots/StatsPlots.jl/pull/356 to *StatsPlots* should solve the problem :)

**Comment by nalimilan** [comment]
So this can be closed now?

**Comment by diegozea** [comment]
Yes, this is solved in *StatsPlots* *0.14.7*
